src/Microsoft.DotNet.XHarness.TestRunners.NUnit/NUnit3XmlOutputWriter.cs

```diff
@@ -0,0 +1,138 @@
+            _xmlWriter.WriteAttributeString("passed", result.PassedTests.ToString());
+            _xmlWriter.WriteAttributeString("failed", result.FailedTests.ToString());
+            _xmlWriter.WriteAttributeString("inconclusive", result.InconclusiveTests.ToString());
+            _xmlWriter.WriteAttributeString("skipped", result.SkippedTests.ToString());
+            _xmlWriter.WriteAttributeString("asserts", result.AssertCount.ToString());
+
+            _xmlWriter.WriteAttributeString("run-date", XmlConvert.ToString(_runStartTime, "yyyy-MM-dd"));
+            _xmlWriter.WriteAttributeString("start-time", XmlConvert.ToString(_runStartTime, "HH:mm:ss"));
+
+            _xmlWriter.WriteAttributeString("random-seed", Randomizer.InitialSeed.ToString());
+
+            WriteEnvironmentElement();
+        }
+
+        void WriteEnvironmentElement()
+        {
+            if (_xmlWriter == null) // should never happen, would mean a programmers error
+                throw new InvalidOperationException("Null writer");
+            _xmlWriter.WriteStartElement("environment");
+
+            var assembly = Assembly.GetExecutingAssembly();
+            AssemblyName assemblyName = AssemblyHelper.GetAssemblyName(assembly);
+            _xmlWriter.WriteAttributeString("nunit-version", assemblyName?.Version?.ToString() ?? "unknown");
+
+            _xmlWriter.WriteAttributeString("clr-version", Environment.Version.ToString());
+            _xmlWriter.WriteAttributeString("os-version", Environment.OSVersion.ToString());
+            _xmlWriter.WriteAttributeString("platform", Environment.OSVersion.Platform.ToString());
+            _xmlWriter.WriteAttributeString("cwd", Environment.CurrentDirectory);
+            _xmlWriter.WriteAttributeString("machine-name", Environment.MachineName);
+            _xmlWriter.WriteAttributeString("user", Environment.UserName);
+            _xmlWriter.WriteAttributeString("user-domain", Environment.UserDomainName);
+            _xmlWriter.WriteAttributeString("culture", CultureInfo.CurrentCulture.ToString());
+            _xmlWriter.WriteAttributeString("uiculture", CultureInfo.CurrentUICulture.ToString());
+
+            _xmlWriter.WriteEndElement();
+        }
+
+        void WriteResultElement(IResultSummary result)
+        {
+            // much simpler than in other writers, we just need to get the child nodes of each of the test-run and write
+            // them. NUnit3 already gave us the xml we need to use
+            foreach (var testRun in result)
+            {
+                for (var i = 0; i < testRun.Result.ChildNodes.Count; i++)
+                {
+                    var node = testRun.Result.ChildNodes[i];
+                    if (node.Name == "environment") continue;
+                    node.WriteTo(_xmlWriter);
+                }
+            }
+        }
+
+        void TerminateXmlFile()
+        {
+            if (_xmlWriter == null) // should never happen, would mean a programmer's error
```

Comment:
Maybe it makes more sense to make `_xmlWriter` a non-nullable field that would throw when not initialized and you wouldn't have to care about these checks?